runtime/protos/runtime.proto

```diff
@@ -17,232 +17,330 @@ option go_package = "github.com/ServiceWeaver/weaver/runtime/protos";
+// weavelet send EnvelopeMsg and WeaveletMsg protocol buffers over these pipes,
+// respectively. We call the part of a deployer that communicates with a
+// weavelet an envelope, hence the name EnvelopeMsg.
+//
+//                 ┌──────────┐   ________   ┌──────────┐
+//                 │          │<─O________)──│          │
+//                 │ ENVELOPE │   ________   │ WEAVELET │
+//                 │          │──O________)─>│          │
+//                 └──────────┘              └──────────┘
+//
+// When an envelope and weavelet first establish their connection, they perform
+// a *handshake*. The envelope sends an EnvelopeInfo, and the weavelet responds
+// with a WeaveletInfo.
+//
+//                   ENVELOPE                  WEAVELET
+//                   ───┬────                  ───┬────
+//                      │ ──────────────────────> │ EnvelopeInfo
+//                      │ <────────────────────── │ WeaveletInfo
+//                      :                         :
+//
+// After the handshake, the envelope and weavelet communicate freely. There are
+// three communication patterns.
+//
+//     1. The envelope can initiate an RPC against the weavelet.
+//     2. The weavelet can initiate an RPC against the envelope.
+//     3. The weavelet can send messages to the envelope (without any replies).
+//
+// Here is an example of these three types of communication.
+//
+//                   ENVELOPE                  WEAVELET
+//                   ───┬────                  ───┬────
+//                   1. │ ──────────────────────> │ UpdateComponentsRequest
+//                      │ <────────────────────── │ UpdateComponentsReply
+//                      │                         │
+//                   2. │ <────────────────────── │ GetListenerAddressRequest
+//                      │ ──────────────────────> │ GetListenerAddressReply
+//                      │                         │
+//                   3. │ <────────────────────── │ LogEntry
+//                      │ <────────────────────── │ TraceSpans
+//                      :                         :
+//
+// Note that these three types of communication can intersperse.
+//
+//                   ENVELOPE                  WEAVELET
+//                   ───┬────                  ───┬────
+//                      │ <────────────────────── │ EnsureComponentRequest
+//                      │ ──────────────────────> │ UpdateComponentsRequest
+//                      │ <────────────────────── │ UpdateComponentsReply
+//                      │ <────────────────────── │ LogEntry
+//                      │ ──────────────────────> │ UpdateRoutingInfoRequest
+//                      │ <────────────────────── │ UpdateRoutingInfoReply
+//                      │ ──────────────────────> │ EnsureComponentReply
+//                      :                         :
+//
+// To avoid deadlocks, a weavelet is guaranteed to process an RPC without
```

Comment:
Do you think it's worth mentioning here about any particular order of these messages?  E.g., what is the sequence of messages a deployer should do (or provide recommendations if no particular order should be enforced).

Or I guess that will go into the deployer blog post.